src/random-seed/random-seed.c

```diff
@@ -306,6 +309,71 @@ static int save_seed_file(
+
+        /* We're doing this opportunistically, so if the seeding dance before didn't manage to initialize the
+         * RNG, there's no point in doing it here. Secondly, getrandom(GRND_NONBLOCK) has been around longer
+         * than EFI seeding anyway, so there's no point in having non-getrandom() fallbacks here. So if this
+         * fails, just return early to cut our losses. */
+        r = getrandom(buffer, sizeof(buffer), GRND_NONBLOCK);
+        if (r < 0 && errno == EAGAIN)
+                return log_debug_errno(-errno, "Random pool not initialized yet, so skipping EFI seed update");
+        else if (r < 0)
+                return log_error_errno(-errno, "Failed to generate random bytes for EFI seed: %m");
+        assert(r == sizeof(buffer));
+        sha256_init_ctx(&hash_state);
+        sha256_process_bytes(&r, sizeof(r), &hash_state);
+        sha256_process_bytes(buffer, r, &hash_state);
+
+        r = find_esp_and_warn(NULL, NULL, /* unprivileged_mode= */ false, &esp_path,
+                              NULL, NULL, NULL, NULL, NULL);
+        if (r == -ENOKEY)
+                return log_debug_errno(r, "Couldn't find any ESP, so not updating ESP random seed.");
+        else if (r < 0)
+                return r; /* find_esp_and_warn() already logged */
+
+        /* Read the existing seed, but if it's not there, return early, since it means bootctl hasn't yet set
+         * up random seeds for the loader. */
+        seed_fd = chase_symlinks_and_open("/loader/random-seed", esp_path, CHASE_PREFIX_ROOT,
+                                          O_RDWR|O_CLOEXEC|O_NOCTTY, NULL);
+        if (seed_fd < 0)
+                return log_debug_errno(-errno, "Failed to open EFI seed path: %m");
+        r = random_seed_size(seed_fd, &len);
+        if (r < 0)
+                return log_error_errno(r, "Failed to determine EFI seed path length: %m");
+        seed_file_bytes = malloc(len);
+        if (!seed_file_bytes)
+                return log_oom();
+        r = loop_read(seed_fd, seed_file_bytes, len, false);
+        if (r < 0)
+                return log_error_errno(r, "Failed to read EFI seed file: %m");
+        sha256_process_bytes(&r, sizeof(r), &hash_state);
+        sha256_process_bytes(seed_file_bytes, r, &hash_state);
+        sha256_finish_ctx(&hash_state, buffer);
+
+        if (lseek(seed_fd, 0, SEEK_SET) < 0)
+                return log_error_errno(-errno, "Failed to seek to beginning of EFI seed file: %m");
+        r = loop_write(seed_fd, buffer, sizeof(buffer), false);
+        if (r < 0)
+                return log_error_errno(r, "Failed to write new EFI seed file: %m");
+        if (ftruncate(seed_fd, sizeof(buffer)) < 0)
+                return log_error_errno(-errno, "Failed to truncate EFI seed file: %m");
+        r = fsync_full(seed_fd);
+        if (r < 0)
+                return log_error_errno(-errno, "Failed to fsync EFI seed file: %m");
+
+        return 0;
```

Comment:
hmm, I think it might be nice to add `log_info("Updated random seed in ESP.")` here, just to report success.

Reply:
Okay.